controllers/operatorscaler_controller.go

```diff
@@ -0,0 +1,334 @@
+//+kubebuilder:rbac:groups=ceph.rook.io,resources=cephclusters,verbs=get;list;watch
+//+kubebuilder:rbac:groups=noobaa.io,resources=noobaas,verbs=get;list;watch
+//+kubebuilder:rbac:groups=monitoring.coreos.com,resources=prometheuses,verbs=get;list;watch
+//+kubebuilder:rbac:groups=monitoring.coreos.com,resources=alertmanagers,verbs=get;list;watch
+//+kubebuilder:rbac:groups=odf.ibm.com,resources=flashsystemclusters,verbs=get;list;watch
+
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.1/pkg/reconcile
+func (r *OperatorScalerReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	r.ctx = ctx
+	r.log = log.FromContext(ctx)
+
+	r.log.Info("starting reconcile")
+
+	if err := r.isOdfDependenciesCsvReady(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileOperators(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	if err := r.reconcileDynamicWatchers(); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.log.Info("reconcile completed successfully")
+	return ctrl.Result{}, nil
+}
+
+func (r *OperatorScalerReconciler) isOdfDependenciesCsvReady() error {
+
+	odfDepsCsv := &oprv1a1.ClusterServiceVersion{}
+
+	err := r.Client.Get(r.ctx, types.NamespacedName{Name: OdfDepsSubscriptionStartingCSV, Namespace: r.OperatorNamespace}, odfDepsCsv)
+	if err != nil {
+		r.log.Error(err, "failed getting odf-deps csv", "csvName", OdfDepsSubscriptionStartingCSV)
+		return err
+	}
+
+	if odfDepsCsv.Status.Phase != oprv1a1.CSVPhaseSucceeded {
+		err = fmt.Errorf("csv %s is not in succeeded state", OdfDepsSubscriptionStartingCSV)
+		r.log.Error(err, "waiting for csv to be in succeeded state")
+		return err
+	}
+
+	return nil
+}
+
+func (r *OperatorScalerReconciler) reconcileOperators() error {
+
+	var returnErr error
+
+	for i := range ResourceMappingList {
+		resourceMapping := ResourceMappingList[i]
+
+		crList := &metav1.PartialObjectMetadataList{}
+		crList.TypeMeta.APIVersion = resourceMapping.ApiVersion
+		crList.TypeMeta.Kind = resourceMapping.Kind
+
+		err := r.Client.List(r.ctx, crList, client.Limit(1))
+		if err != nil {
+			if !meta.IsNoMatchError(err) {
+				msg := fmt.Sprintf("failed listing %s", resourceMapping.Kind)
+				r.log.Error(err, msg)
+				multierr.AppendInto(&returnErr, err)
+			}
+
+		} else if len(crList.Items) >= 0 {
+
+			for _, pkgName := range resourceMapping.PkgNames {
+				key := fmt.Sprintf(CsvLabelKey, pkgName, r.OperatorNamespace)
+
+				csvList := &oprv1a1.ClusterServiceVersionList{}
+				err = r.Client.List(
+					r.ctx, csvList,
+					client.InNamespace(r.OperatorNamespace),
+					client.MatchingLabels(map[string]string{key: ""}),
+				)
+				if err != nil {
+					r.log.Error(err, "failed listing csvs with label", "label", key)
+					multierr.AppendInto(&returnErr, err)
+				} else {
+					err = r.updateCsvDeplymentsReplicas(csvList.Items...)
```

Comment:
As asked in the last review (https://github.com/red-hat-storage/odf-operator/pull/529#discussion_r1983419507), we should be looping over the CSVs here then sending each CSV into the func one by one.

Reply:
I have replied there.

Reply:
Added a loop here